Script manipulation via lua

Edru2 opened the pull request:
#### Brief overview of PR changes/additions
New functions:
**permScript**(name, parent, luacode) -- works the same way as other perm functions
**enableScript**("script-name") -- enables all scripts found with the name "script-name"
**disableScript**("script-name")-- disables all scripts found with the name "script-name"
**getScript**("script-name", [occurrence]) -- returns the luacode of the n occurrence of  "script-name" as string (if occurence is not set it is 1 which is the first occurrence)
**setScript**("scriptname", luacode, [occurrence]) -- set the luacode of the n occurrence of "scriptname" (overwrites old code, if occurrence is not set it is 1 which is the first occurrence
**appendScript**("script-name", luacode, [occurrence]) -- appends luacode to n occurrence of "scriptname" (if occurrence is not set it is 1 which is the first occurrence)
additionally **permGroup** accepts the type "*script*" now.
#### Motivation for adding to Mudlet
Mudlet was lacking this abilities by itself.
Script manipulation was only possible in a hacky way.
https://github.com/Mudlet/Mudlet/issues/2897
#### Other info (issues closed, discussion etc)
The new permGroup and permScript functions have the same limitations as the old ones (for timers, alias...).

It is possible to get/set a script even with the same name through the value occurrence.
To know how many occurrences of the same script are there it's possible to use: **exists("myscriptname", "script")**

Found an issue (and fixed it) in exists and isActive for scripts, they were only finding scripts which were located in the root node.

Edru2 committed:
initial commit

permScript
getScript
setScript

Edru2 committed:
new Functions enable/disable Script permGroup for script

Edru2 committed:
Merge branch 'development' into string-to-xml-script

demonnic commented:
This all works, except that it won't let you get or set the script on a group, and they can contain lua code.

busted tests pass

Edru2 committed:
added position of occurence

Edru2 committed:
fix appendScript and setScript

Edru2 commented:
> This all works, except that it won't let you get or set the script on a group, and they can contain lua code.

It was made this way because I didn't know that a script in a group is a thing. (changed it)

Now it is possible to also get and set the script in a group.

I got over the limitation of only able to set, get the first occurence of a script by adding the parameter occurence.

So now it is possible to have the same Group and Script name and get/set to the script only.

vadi2 commented on src/mudlet-lua/lua/Other.lua:
```diff
@@ -180,13 +180,14 @@ end
   assert(type(name) == "string", "appendScript: bad argument #2 type (lua code as string expected, got "..type(luaCode).."!)")
-  if not getScript(name) then
+  if not getScript(name, pos) then
     return nil, 'script "'..name..'" not found'
```
Make sure to include the position you searched for too, in case that was passed as a variable - and the developer isn't super sure of what position it searched for.

Edru2 replied:
Very strange. I did this already in my last commit. You seem to reference one commit earlier.

Edru2 replied:
Aah my fault... Now I understand what you mean

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -8295,6 +8345,116 @@ int TLuaInterpreter::permAlias(lua_State* L)
+}
+
+// Documentation: https://wiki.mudlet.org/w/Manual:Lua_Functions#setScript
+int TLuaInterpreter::setScript(lua_State* L)
```
This doesn't seem to work for new or existing scripts - just returns `-1`.

Edru2 replied:
Same here. 
If the script is not compiling successfully it will not change the existing script.
It should return also a string with the error message?

vadi2 replied:
ohh for sure, and a false instead of -1

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -8295,6 +8345,116 @@ int TLuaInterpreter::permAlias(lua_State* L)
+
+
+// Documentation: https://wiki.mudlet.org/w/Manual:Lua_Functions#permScript
+int TLuaInterpreter::permScript(lua_State* L)
```
Same as `setScript()`, am I doing something wrong?

Edru2 replied:
Is there an error message?
Probably the script is not compiling.
If it's not compiling it won't create the script (same behaviour as for permTimer/permAlias...)

Edru2 committed:
false if script is not compiled succesfully

return false if script not compiled
and use of QVector.value to check if the value is really there
to avoid crashes of Mudlet and other strange behaviour

SlySven commented:
:thinking: With regard to returning an error message should a script fail to compile - are we sure we are doing that at present for ALL the other Mudlet item types (aliases, buttons/menus/toolbar {though only the FIRST of those can have an executable Lua script IIRC}, keybindings, timers and all the different flavors of trigger} and for aliases and triggers with PCREs also checking that those compile as well?

Not an issue specifically for this PR but something it might be worth confirming...

vadi2 commented:
I don't think so - but now we realise it's a good idea...

Edru2 commented:
You're right. Didn't check them all but it's done for perm and temp timers at least.

SlySven commented:
> ... Didn't check them all but it's done for perm and temp timers at least.

Yeah - 'cos I had a go at fixing them a while back - am looking now (and raising an issue with a checklist) and I see that many of the Lua API calls just return a ID number - which might be `-1` on error but do NOT pass an error message as well. :cry:

vadi2 commented:
Looks like a bit of refactoring for better quality is needed, it's including the html intended for the ui into the response:

![image](https://user-images.githubusercontent.com/110988/77829485-ac31e800-7122-11ea-985d-2602ec4a1125.png)

SlySven commented:
I mention that in that new issue I mentioned above. It is the subject of Issue #3039 .

SlySven commented on src/dlgTriggerEditor.cpp:
```diff
@@ -4352,6 +4352,25 @@ void dlgTriggerEditor::saveAction()
+    }
+
+    TScript* pT = mpHost->getScriptUnit()->getScript(scriptID);
+    QString scriptCode = pT->getScript();
```
:warning: Coverity is reporting that there is no test for `pT` being a `nullptr` before it is dereferenced here with a reference: **CID 1491992:  Null pointer dereferences  (NULL_RETURNS)**
